pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/RealtimeSegmentAssignmentStrategy.java

```diff
@@ -0,0 +1,155 @@
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.apache.pinot.common.config.TableConfig;
+import org.apache.pinot.common.utils.LLCSegmentName;
+import org.apache.pinot.controller.helix.core.assignment.InstancePartitions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Segment assignment strategy for LLC real-time segments.
+ * <p>It is very similar to replica-group based segment assignment with the following differences:
+ * <ul>
+ *   <li>1. Each partition always contains one server</li>
+ *   <li>2. In addition to the ONLINE segments, there are also CONSUMING segments to be assigned</li>
+ * </ul>
+ * <p>
+ *   Since each partition contains only one server (in one replica), we can directly assign or rebalance segments to the
+ *   servers based on the partition id.
+ * <p>
+ *   The real-time segment assignment does not minimize segment moves because the server is fixed for each partition in
+ *   each replica. The instance assignment is responsible for keeping minimum changes to the instance partitions to
+ *   reduce the number of segments need to be moved.
+ */
+public class RealtimeSegmentAssignmentStrategy implements SegmentAssignmentStrategy {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RealtimeSegmentAssignmentStrategy.class);
+
+  private String _tableNameWithType;
+  private int _replication;
+
+  @Override
+  public void init(ZkHelixPropertyStore<ZNRecord> propertyStore, TableConfig tableConfig) {
+    _tableNameWithType = tableConfig.getTableName();
+    _replication = tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+
+    LOGGER.info("Initialized RealtimeSegmentAssignmentStrategy for table: {} with replication: {}", _tableNameWithType,
+        _replication);
+  }
+
+  @Override
+  public List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
+      InstancePartitions instancePartitions) {
+    Preconditions.checkState(instancePartitions.getNumPartitions() == 1,
+        "The instance partitions should contain only 1 partition");
+
+    int numReplicas = getNumReplicas(instancePartitions);
+    int partitionId = new LLCSegmentName(segmentName).getPartitionId();
+    List<String> instancesAssigned = new ArrayList<>(numReplicas);
+    for (int replicaId = 0; replicaId < numReplicas; replicaId++) {
+      instancesAssigned.add(getInstance(instancePartitions, partitionId, replicaId));
+    }
+    LOGGER.info("Assigned segment: {} with partition id: {} to instances: {} for table: {}", segmentName, partitionId,
+        instancesAssigned, _tableNameWithType);
+    return instancesAssigned;
+  }
+
+  @Override
+  public Map<String, Map<String, String>> rebalanceTable(Map<String, Map<String, String>> currentAssignment,
```

Comment:
should include a flag whether we need to rebalance consuming segments or not?

Reply:
Discussed with @npawar. In order to benefit from replica-group routing (query less servers), we need to always rebalance consuming segments if partition for an instance changed (should be rare).